### ISSUE TITLE
please move "math_eigen.h" to "src" and accept a pull request from "mrcdr"

### ISSUE THREAD

**Issue by jewettaij** [lammps/lammps#2321]
**Summary**

1) Please accept a future pull request from git user "mrcdr" so he can get credit for his work.  (It will be empty.)

2) Please also consider moving "math_eigen.h" into the main LAMMPS "src" directory.  

**LAMMPS Version and Platform**

This is not specific to any version of LAMMPS

**Details**

Jacob, mrcdr, and I worked on an "rmsd" feature together that required calculating eigenvectors of a 4x4 matrix.  I'm thrilled to see that this feature was accepted and is now part of LAMMPS (https://github.com/lammps/lammps/pull/2314)

However in order to secure mrcdr's help, I promised him that we would credit him on github for his contribution, and I think it was reasonably important to him.  But it seems Jacob's pull request accidentally erased the commit history of our files.  (It's easy to do.  Git is a pain.)  Is there a way would could fix this?  

### requests

1) If "mrcdr" makes a pull request that adds a single blank line to the end of the "math_eigen.h" file, would you be willing to accept it?  This way, his name is added to the list of LAMMPS contributors on github.

2) I would also like to move the "math_eigen.h" file into the main LAMMPS "src" directory.

Both Yuya ("mrcdr") and I labored hard to write and test a reasonably general (and very compact) eigen decomposition library which is compatible with the GPL and LGPL licenses. (If all we wanted to do was add a quick feature to USER-REACTION, we would have wrote this code differently.)  The code in our file ("math_eigen.h") has been commented, documented, and extensively tested using automated tests:
https://github.com/jewettaij/jacobi_pd
https://github.com/mrcdr/lambda-lanczos
I really did not want to write this code.  It took months of our time.  You may not be aware of it, but anyone who needs to use eigen-solvers with LAMMPS really struggles with this issue.
https://lammps.sandia.gov/threads/msg84021.html

That's why I want to move it to the "src" directory: To help other future LAMMPS developers find this code. 

*(It also would help to solve a legal problem. Currently LAMMPS contains code, in "math_extra.h", which was taken from Numerical Recipes in C, in violation of the Numerical Recipes license. The "Jacobi" class in "math_eigen.h" contains a public-domain version of this algorithm which is legal and also has more general applicability.)*

Axel suggested asking code contributors just use Eigen3 instead.  I'm fine with that, as long as both build systems (cmake and make) can be configured to automatically download and install Eigen3 when needed.  (Requiring end-users to download Eigen3 manually puts the feature developer at a disadvantage.  Fewer people will use that feature and cite the method.)  Moving "math_eigen.h" into "src" seems easier and is harmless.

I'm sorry to sound so grumpy over this.  I'm not seeking credit or glory.  I just want to save future developers from going through what we went through.  (It was really a pain.)

Thanks for considering this.

**Comment by jewettaij** [comment]
After having written this, I am embarrassed.  It sounds like I am taking out my frustration on the LAMMPS developers.  That is really not the tone I meant to use.  And I am not frustrated at all.  I'm really happy that the "math_eigen.h" file was accepted.  I just want to reduce pain and suffering for future developers.

Either way, if you get a (mostly blank) pull request from "mrcdr", please accept it.  (He will probably move the "math_eigen.h" file, but you can move it back.)

**Comment by akohlmey** [comment]
> However in order to secure mrcdr's help, I promised him that we would credit him on github for his contribution, and I think it was reasonably important to him. But it seems Jacob's pull request accidentally erased the commit history of our files. (It's easy to do. Git is a pain.) Is there a way would could fix this?

The git history would be included if there was a pull request (with the commit history included) that adds the file directly to the src folder, but then removes the one from the src/USER-REACTION folder. Its history was lost by not including it into USER-REACTION with a pull request to @jrgissing's repo, but through copying the file into it.

For it to be moved to src, it would have to more closely comply with the conventions for LAMMPS code.
Due to it being a header-only library, it is a bit complicated, but it can be made to work, if it is not included in header with style class definitions.  One (cosmetic) change required would the the guard to protect against multiple inclusions would have to be renamed to LMP_MATH_EIGEN_H.
 
> Both Yuya ("mrcdr") and I labored hard to write and test a reasonably general (and very compact) eigen decomposition library which is compatible with the GPL and LGPL licenses. (If all we wanted to do was add a quick feature to USER-REACTION, we would have wrote this code differently.) The code in our file ("math_eigen.h") has been commented, documented, and extensively tested using automated tests:
> https://github.com/jewettaij/jacobi_pd
> https://github.com/mrcdr/lambda-lanczos
> I really did not want to write this code. It took months of our time. You may not be aware of it, but anyone who needs to use eigen-solvers with LAMMPS really struggles with this issue.
> https://lammps.sandia.gov/threads/msg84021.html

If you have it documented and some tester code, **and** you want it to be used more generally in LAMMPS, then:
- it's documentation and some simple usage example(s) should be included into the new programmer guide section of the manual (see PR #2319 and #2320) as its own subsection, similar to the "LAMMPS utility functions" subsection.
We have just added support for including doxygen decorations into sphinx via the "breathe" extension to sphinx
- the tester code should be adapted to using googletest and added to unittest/utils/test_math_eigen.cpp
- a pull request would be highly welcome that replaces the current usage of MathExtra::jacobi with this new code and removes the old version. doing this replacement is a goal for the upcoming stable release. we have some alternate suggested code for this that had some stability issues and thus is currently on hold pending further tests and possible improvements. but your contribution could make this task obsolete (assuming there are no similar issues with this new code, but in that case the issues might need to be addressed somehow in the calling code).

> That's why I want to move it to the "src" directory: To help other future LAMMPS developers find this code.

That doesn't really help that much on its own. This is why the programmer guide is being integrated to the manual and it is including documentation for recommended utility functions and classes (example: am currently working on similar docs for the new Tokenizer classes to parse strings into words and numbers, that are supposed to replace the more convoluted parsers using strtok()).

> _(It also would help to solve a legal problem. Currently LAMMPS contains code, in "math_extra.h", which was taken from Numerical Recipes in C, in violation of the Numerical Recipes license. The "Jacobi" class in "math_eigen.h" contains a public-domain version of this algorithm which is legal and also has more general applicability.)_

Nod. We are still aware of this issue and it needs to be addressed soon.

> Axel suggested asking code contributors just use Eigen3 instead. I'm fine with that, as long as both build systems (cmake and make) can be configured to automatically download and install Eigen3 when needed. (Requiring end-users to download Eigen3 manually puts the feature developer at a disadvantage. Fewer people will use that feature and cite the method.) Moving "math_eigen.h" into "src" seems easier and is harmless.

Both CMake and conventional make can download Eigen3 transparently (except for a little bug that has crept into src/Makefile a couple months ago and should be fixed in one of the upcoming pull requests). Once that bug is fixed, Eigen3 can be downloaded/installed in conventional make via `make lib-smd args=-b`. It already works by going to `lib/smd` and typing `python Install.py -b` (if you have a recent checkout, since they moved their hosting provider recently).
But since we don't want this download (or internet access at compile time) be a requirement for compiling the core LAMMPS code, having `math_eigen.h` in src is indeed a viable alternative and the reason why Eigen3 is not a required dependency of LAMMPS.

> I'm sorry to sound so grumpy over this. I'm not seeking credit or glory. I just want to save future developers from going through what we went through. (It was really a pain.)

Making your life miserable is somehow part of developing software, especially for open source scientific software. Being able to deal with the resulting frustration is part of the challenge. However trying to prevent others from repeating that frustration is rarely having a significant effect as 1) somehow people seem to be bound to discover the ways to avoid frustration only after the fact and 2) bypassing one source of frustration often just leads to getting stuck and frustrated with something else. The balance between frustration and gratification appears to be always tilted toward frustration.

**Comment by jrgissing** [comment]
sorry about the lost authorship issues, and thanks for bringing it up (it is no trivial issue). it will also be important if/when other people modify those files, so your name should be attached to every line of those files

i eventually discovered the official 'git mv' command when moving around other people's files, but i may have been too late. however, therefore it should be equally easy to restore the correct authorship. for example, in addition to the lammps-specific corrections Axel mentions above, perhaps you can do a commit that deletes those files in user-reaction, and another commit that puts them back. or to wherever you'd like them

**Comment by jewettaij** [comment]
> sorry about the lost authorship issues, and thanks for bringing it up (it is no trivial issue). it will also be important if/when other people modify those files, so your name should be attached to every line of those files

Don't worry about this by the way.  I forget to use "git mv" all the time.

> i eventually discovered the official 'git mv' command when moving around other people's files, but i may have been too late. however, therefore it should be equally easy to restore the correct authorship. for example, in addition to the lammps-specific corrections Axel mentions above, perhaps you can do a commit that deletes those files in user-reaction, and another commit that puts them back. or to wherever you'd like them

That's a great idea.  We might do that with "math_eigen.hpp".  (Although I won't do this with any bigger files.)

No onward to reply to Axel...

**Comment by jewettaij** [comment]
> For it to be moved to src, it would have to more closely comply with the conventions for LAMMPS code.

Great!  (relieved)  Thanks for considering this.

> Due to it being a header-only library, it is a bit complicated, but it can be made to work, if it is not included in header with style class definitions.

Not sure.  Other than a few lines of fancy [C++ type manipulation magic](https://github.com/lammps/lammps/blob/3be064748d24efa715ff9c0d27cbbe1267f551a2/src/USER-REACTION/math_eigen.h#L85) that Yuya added, the file is relatively straightforward.  It defines a couple small inlined functions, and 3 classes: Jacobi, LambdaLanczos, PEigenDense (which is a convenience wrapper for LambdaLanczos).

The contents of "math_eigen.h" are currently enclosed in a namespace ("MathEigen").  We can either keep it there, or move it to "MathExtra" (or "utils"?) namespaces.  (There are some functions in "math_eigen.h" named "Alloc2D()" and "Dealloc2D()", "inner_prod()", "normalize()" and "scalar_mul()".  Fortunately, these don't seem to clash with any existing functions in LAMMPS yet.  Alloc2D() was intended for allocating small (eg. 3x3, or 4x4) 2D arrays.  It is similar to the 2D "create()" function in "memory.h", but you don't have to give each 2D array a name.  Unless you suggest otherwise, I'll keep all this stuff in the "MathEigen" namespace.)

> One (cosmetic) change required would the the guard to protect against multiple inclusions would have to be renamed to LMP_MATH_EIGEN_H.

Sounds good to me.

> If you have it documented and some tester code, **and** you want it to be used more generally in LAMMPS, then:
> 
> * it's documentation and some simple usage example(s) should be included into the new programmer guide section of the manual (see PR #2319 and #2320) as its own subsection, similar to the "LAMMPS utility functions" subsection.
>   We have just added support for including doxygen decorations into sphinx via the "breathe" extension to sphinx

That's great, by the way.  It sounds like you are making some important, fundamental infrastructure changes to LAMMPS these days.  The ability for users to modify the LAMMPS code is one of its greatest strengths so I'm glad the documentation there is getting some love.

As far as "math_eigen.h" goes, most of documentation and usage examples already exists as doxygen comments in the "math_eigen.h" file.  While I am trying to use proper doxygen syntax, I'd be surprised if I didn't make any mistakes.  How can I check if doxygen (and sphynx+breath) understands the doxygen comments that I added?

> * the tester code should be adapted to using googletest and added to unittest/utils/test_math_eigen.cpp
I can make a test_math_eigen.cpp file and take a stab at converting these tests to googletest.  It will take a couple days to figure out how Yuya tests his code and get all this working, but I can do it.

> * a pull request would be highly welcome that replaces the current usage of MathExtra::jacobi with this new code and removes the old version. doing this replacement is a goal for the upcoming stable release. we have some alternate suggested code for this that had some stability issues and thus is currently on hold pending further tests and possible improvements. but your contribution could make this task obsolete (assuming there are no similar issues with this new code, but in that case the issues might need to be addressed somehow in the calling code).

I'll start with this, then.  This is easy for me to do.

Do you want me to submit a pull request once I've got this done?

Or do you want me to wait until I've got the googletest and doxygen stuff working too?

> > That's why I want to move it to the "src" directory: To help other future LAMMPS developers find this code.
> 
> That doesn't really help that much on its own. This is why the programmer guide is being integrated to the manual and it is including documentation for recommended utility functions and classes (example: am currently working on similar docs for the new Tokenizer classes to parse strings into words and numbers, that are supposed to replace the more convoluted parsers using strtok()).

Please take a quick look at the doxygen comments for the [Jacobi](https://github.com/lammps/lammps/blob/3be064748d24efa715ff9c0d27cbbe1267f551a2/src/USER-REACTION/math_eigen.h#L125) and [LambdaLanczos](https://github.com/lammps/lammps/blob/3be064748d24efa715ff9c0d27cbbe1267f551a2/src/USER-REACTION/math_eigen.h#L266) classes, and let me know if this is what you need.

It's not perfect.  I can spot some issues.  (I forgot to enclose the Jacobi usage example in a @code, ...@endcode block.  I forgot to provide a usage example for the "PEigenDense" class.)  I will fix these issues.

Please let me know if I'm on the right track with the way I'm documenting the code so far.

> > _(It also would help to solve a legal problem. Currently LAMMPS contains code, in "math_extra.h", which was taken from Numerical Recipes in C, in violation of the Numerical Recipes license. The "Jacobi" class in "math_eigen.h" contains a public-domain version of this algorithm which is legal and also has more general applicability.)_
> 
> Nod. We are still aware of this issue and it needs to be addressed soon.

> Both CMake and conventional make can download Eigen3 transparently (except for a little bug that has crept into src/Makefile a couple months ago and should be fixed in one of the upcoming pull requests). Once that bug is fixed, Eigen3 can be downloaded/installed in conventional make via `make lib-smd args=-b`. It already works by going to `lib/smd` and typing `python Install.py -b` (if you have a recent checkout, since they moved their hosting provider recently).
> But since we don't want this download (or internet access at compile time) be a requirement for compiling the core LAMMPS code, having `math_eigen.h` in src is indeed a viable alternative and the reason why Eigen3 is not a required dependency of LAMMPS.

Good.  I agree.

> > I'm sorry to sound so grumpy over this. I'm not seeking credit or glory. I just want to save future developers from going through what we went through. (It was really a pain.)
> 
> Making your life miserable is somehow part of developing software, especially for open source scientific software. Being able to deal with the resulting frustration is part of the challenge. However trying to prevent others from repeating that frustration is rarely having a significant effect as 1) somehow people seem to be bound to discover the ways to avoid frustration only after the fact and 2) bypassing one source of frustration often just leads to getting stuck and frustrated with something else. The balance between frustration and gratification appears to be always tilted toward frustration.

I think you experience more of this than the average person who tinkers with the LAMMPS code.
Thank you for being receptive to moving "math_eigen.h", and for all your unsung, and critically important work.

**Comment by akohlmey** [comment]
@jewettaij I will try to address your questions and make some suggestion in one block rather than responding to individual statements and sections.

First off, please update your LAMMPS checkout to the latest state of the master branch. we just merged a whole bunch of stuff related to including doxygen and utility function documentation. You can go to https://docs.lammps.org/ and view the current status of the manual. Please have a particular look at the part called "developer guide" https://docs.lammps.org/pg_developer.html and especially the bottom parts of it, where several utility functions in the utils namespace and the tokenizer and file reader classes are documented. This is where I would like the documentation for this functionality go.

I would prefer to leave it in its MathEigen namespace. It is a separate entity and the namespace represents that nicely.
It is not that good a fix with utils:: since those are usually smaller in scope.

We have opted to go for a mixed documentation scheme instead of going all-doxygen. So we process only selected files with doxygen into xml files and then use the "breathe" extension to sphinx to import the low level API description. This is combined with text and examples written directly in .rst. This means, high-level docs are in the doc folder but the API description is left with the source code, so that it can be most easily kept in sync. The doxygen directives are mostly written in the Qt style, with a few javadoc like comments thrown in where they are more convenient to use.

The concern about header-only implementations is that they pull in other headers and we want to keep this to the minimum for all files that have style classes in them, since those will be loaded all at the same time by the style_xxx.h header files. In the past we repeatedly had to deal with hard to debug and resolve issues because of that. For contributed code we thus often ask people to PIMPLify their code, i.e. put all references to implementation specific data structures into a struct instead of the class directly and then put a pointer to that struct into the class itself, which can be handled in the header with a forward declaration. the specific headers are then only included in the implementation, where there are no conflicts with other classes and their headers and definitions. USER-SCAFACOS and USER-PLUMED are examples for that.

Another issue of a template library is that it can blow up the executable with redundant instantiations. Especially in this case, there is little need for that, since we would only be using doubles.That doesn't mean that you need to rewrite the entire code.
You could just put the essential APIs (as templates) into the header and the rest of the code into a matching .cpp file and then do an explicit instantiation (or multiple, if there is a need). If you need an example for that, the simplest is probably the recent change where `Force::bounds()` and `Force::boundsbig()`  were combined into `utils::bounds()`. You will see only the prototype in utils.h, and then the implementation  in utils.cpp and then 3 explicit instantiations that are required for all uses. So ideally we would end up with a "lean-and-mean" header that has only the absolute minimum of interfaces and include statements and everything else could move to the .cpp file.

We can do the integration in multiple steps. Some of the unit and regression tests will likely need to be updated when we switch the implement for existing functionality.

**Comment by akohlmey** [comment]
Some more info on the doxygen integration. To have a new file processed by doxygen, you need to add a new entry to doc/doxygen/Doxyfile.in. doxygen will be automatically run when you type "make html" or "make pdf".
then you add a section to the doc/src/pg_developer.rst file (no worries if it seems large, we will break it into pieces later)
and then you can add statements like `.. doxygenclass::` or `.. doxygenfunction::` to import the API description from the source.
the file `doc/doxygen-warn.log` will contain warnings from the doxygen processing (currently none). After doing "make html" you should also do "make spelling" to flag typos and non-words that need to be added to the list of false positives.

**Comment by jewettaij** [comment]
## Question: May I create a global variable ?
The main issue I am currently running into is the need to create a temporary array to store intermediate calculations.  The existing "jacobi()" in "math_extra.h" gets around this issue by 1) restricting itself to 3x3 matricies, 2) not attempting to preserve the contents of the matrix it is operating on.  ("jacobi()" destroys the contents of the matrix argument as it runs).  In my version, Jacobi is a class which needs to be instantiated before use.  The constructor takes as an argument, the size of the matrices it will operate on.  This works fine, however it creates some small overhead because before you calculate the eigenvalues, a new instance of Jacobi() must be created (a new 3x3 matrix must be allocated).

## Solution method 1: Create a global variable of type "Jacobi"

```
  MathEigen::Jacobi<double,double*,double(*)[3]> g_ecalc3(3);
  int Jacobi3(double const mat[3][3],// the 3x3 matrix you wish to diagonalize
              double *eval,          // store the eigenvalues here
              double evec[3][3])     // store the eigenvectors here (in cols)
  {
    int ierror = g_ecalc.Diagonalize(mat, eval, evec);
    // details: transpose the evec matrix for compatibility with jacobi()  (omitted...)
    return ierror;
  }
}
```
The nice thing about this approach is that the "Jacobi3()" function is interchangeable with the old "jacobi()" function.  No modification to the existing LAMMPS code needs to be made other than a change in name "jacobi("->"Jacobi3(".  (The "math_eigen.h" file must be included in the .cpp files, but not in the .h files.)

### Solution method 2:
Every LAMMPS class which needs to calculate eigenalues has its class declaration modified so that it has an instance of the Jacobi class as a private data member.  In other words, classes like FixRigid will have a private data member (typically named "ecalc") which is of type "Jacobi" and replaces the "g_ecalc" global variable in the code above.  They will use ecalc.Diagonalize() to find eigenvalues and eigenvectors, instead of Jacobi3().

I prefer not to use this approach, because in this case, all of the header files for these classes: *("group.h", "molecule.h", "compute_omega_chunk.h", "atom_vec_tri.h", "fix_rigid.h", "fix_rigid_small.h", "fix_poems.h", "compute_omega_chunk.h", "body_nparticle.h", "body_rounded_polygon.h", "body_rounded_polyhedron.h", "compute_gyration_shape.h", "compute_gyration_shape_chunk.h")*

...would need to contain this line (in order to define the "Jacobi" class):
```
include "math_eigen.h"
```

If I hear no objections, I plan to use method 1.
-Andrew

**Comment by jewettaij** [comment]
I can think of an uglier solution which is thread safe, but I'll wait until you reject my current proposal before bringing it up.

**Comment by jewettaij** [comment]
Too much of me thinking out loud on my part. (The original version of this comment was a long rant about thread safety.) Sorry about this. I have made a decision. I'll send you my pull request soon, and then we can debate the approach I chose.

**Comment by jewettaij** [comment]
Thanks Axel for filling in the missing bits and taking this all the way to completion.
I will email Yuya and let him know that we are ready for an empty pull request from him.
